File: code/ch07/java/HumbleView.java
// §7.3.3 Keep Views Simple — fat view vs. humble view over a view-model
// (clinic scheduler: invoice list with an overdue badge).
// Compact source file (JDK 25+): run with `java -ea HumbleView.java`.
record Invoice(String patientName, LocalDate sentOn, boolean paid) {}

// --- Before: fat view --------------------------------------------------------
class InvoiceWidget {
  String render(Invoice invoice, LocalDate today) {
    String text = invoice.patientName();
    if (!invoice.paid() && ChronoUnit.DAYS.between(invoice.sentOn(), today) > 30) {
      text += " — OVERDUE";              // a business rule, trapped on screen
    }
    return text;
  }
}

// --- After: humble view ------------------------------------------------------
// the rule, extracted where tests reach it
boolean isOverdue(Invoice invoice, LocalDate today) {
  return !invoice.paid() && ChronoUnit.DAYS.between(invoice.sentOn(), today) > 30;
}

record InvoiceViewModel(String text) {}

InvoiceViewModel invoiceViewModel(Invoice invoice, LocalDate today) {
  String badge = isOverdue(invoice, today) ? " — OVERDUE" : "";
  return new InvoiceViewModel(invoice.patientName() + badge);
}

class HumbleInvoiceWidget {              // the book shows this as InvoiceWidget,
  String render(InvoiceViewModel vm) {   // rewritten; renamed so both versions
    return vm.text();                    // run in one file
  }
}

void testUnpaid31DaysIsOverdue() {
  assert isOverdue(new Invoice("Ana", LocalDate.of(2026, 6, 1), false),
      LocalDate.of(2026, 7, 2));
}

void main() {
  var ana = new Invoice("Ana", LocalDate.of(2026, 6, 1), false);
  var today = LocalDate.of(2026, 7, 2);
  var fat = new InvoiceWidget().render(ana, today);
  assert fat.equals("Ana — OVERDUE");

  var vm = invoiceViewModel(ana, today);
  assert new HumbleInvoiceWidget().render(vm).equals(fat);
  assert !isOverdue(new Invoice("Ben", LocalDate.of(2026, 6, 15), false), today);
  assert !isOverdue(new Invoice("Cho", LocalDate.of(2026, 5, 1), true), today);
  testUnpaid31DaysIsOverdue();
  IO.println("fat view and humble view render identically; rule tested directly");
}

File: code/ch07/python/humble_view.py
# §7.3.3 Keep Views Simple — fat view vs. humble view over a view-model
# (clinic scheduler: invoice list with an overdue badge).
from dataclasses import dataclass
from datetime import date

@dataclass
class Invoice:
  patient_name: str
  sent_on: date
  paid: bool

# --- Before: fat view -------------------------------------------------------
class InvoiceWidget:
  def render(self, invoice, today):
    text = invoice.patient_name
    if not invoice.paid and (today - invoice.sent_on).days > 30:
      text += " — OVERDUE"         # a business rule, trapped on screen
    return text

fat = InvoiceWidget().render(Invoice("Ana", date(2026, 6, 1), False), date(2026, 7, 2))
assert fat == "Ana — OVERDUE"

# --- After: humble view -----------------------------------------------------
def is_overdue(invoice, today):          # the rule, extracted where tests reach it
  return not invoice.paid and (today - invoice.sent_on).days > 30

def invoice_view_model(invoice, today):
  badge = " — OVERDUE" if is_overdue(invoice, today) else ""
  return {"text": invoice.patient_name + badge}

class HumbleInvoiceWidget:               # the book shows this as InvoiceWidget,
  def render(self, vm):                # rewritten; renamed so both versions
    return vm["text"]                # run in one file

def test_unpaid_31_days_is_overdue():
  assert is_overdue(Invoice("Ana", sent_on=date(2026, 6, 1), paid=False), date(2026, 7, 2))

vm = invoice_view_model(Invoice("Ana", date(2026, 6, 1), False), date(2026, 7, 2))
assert HumbleInvoiceWidget().render(vm) == "Ana — OVERDUE"
assert not is_overdue(Invoice("Ben", date(2026, 6, 15), False), date(2026, 7, 2))
assert not is_overdue(Invoice("Cho", date(2026, 5, 1), True), date(2026, 7, 2))
test_unpaid_31_days_is_overdue()
print("fat view and humble view render identically; rule tested directly")
